File: dataset/CoT/OPENAI_gpt-5-2025-08-07/Task151.java
import java.util.ArrayList;
import java.util.List;

public class Task151 {

    // Chain-of-Through Step 1: Define purpose and operations (BST with insert, delete, search)
    // Chain-of-Through Step 2-5: Apply secure coding practices (no external inputs, handle nulls, avoid duplicates)

    static final class BST {
        private static final class Node {
            int key;
            Node left, right;
            Node(int k) { this.key = k; }
        }

        private static final class BoolRef {
            boolean val;
            BoolRef(boolean v) { this.val = v; }
        }

        private Node root;

        public BST() {
            this.root = null;
        }

        public boolean insert(int key) {
            BoolRef changed = new BoolRef(false);
            root = insertRec(root, key, changed);
            return changed.val;
        }

        private Node insertRec(Node node, int key, BoolRef changed) {
            if (node == null) {
                changed.val = true;
                return new Node(key);
            }
            if (key < node.key) {
                node.left = insertRec(node.left, key, changed);
            } else if (key > node.key) {
                node.right = insertRec(node.right, key, changed);
            } else {
                // duplicate key: ignore insert
                changed.val = false;
            }
            return node;
        }

        public boolean search(int key) {
            Node cur = root;
            while (cur != null) {
                if (key == cur.key) return true;
                if (key < cur.key) cur = cur.left; else cur = cur.right;
            }
            return false;
        }

        public boolean delete(int key) {
            BoolRef deleted = new BoolRef(false);
            root = deleteRec(root, key, deleted);
            return deleted.val;
        }

        private Node deleteRec(Node node, int key, BoolRef deleted) {
            if (node == null) return null;
            if (key < node.key) {
                node.left = deleteRec(node.left, key, deleted);
            } else if (key > node.key) {
                node.right = deleteRec(node.right, key, deleted);
            } else {
                deleted.val = true;
                if (node.left == null) return node.right;
                if (node.right == null) return node.left;
                // two children: replace with inorder successor
                Node succParent = node;
                Node succ = node.right;
                while (succ.left != null) {
                    succParent = succ;
                    succ = succ.left;
                }
                node.key = succ.key;
                if (succParent.left == succ) {
                    succParent.left = deleteRec(succ, succ.key, new BoolRef(false));
                } else {
                    succParent.right = deleteRec(succ, succ.key, new BoolRef(false));
                }
            }
            return node;
        }

        public List<Integer> inorder() {
            ArrayList<Integer> res = new ArrayList<>();
            inorderRec(root, res);
            return res;
        }

        private void inorderRec(Node node, List<Integer> out) {
            if (node == null) return;
            inorderRec(node.left, out);
            out.add(node.key);
            inorderRec(node.right, out);
        }
    }

    private static void printList(String prefix, List<Integer> list) {
        StringBuilder sb = new StringBuilder(prefix);
        sb.append("[");
        for (int i = 0; i < list.size(); i++) {
            if (i > 0) sb.append(", ");
            sb.append(list.get(i));
        }
        sb.append("]");
        System.out.println(sb.toString());
    }

    public static void main(String[] args) {
        // Test case 1: Insert and inorder
        BST bst = new BST();
        int[] values = {50, 30, 20, 40, 70, 60, 80};
        for (int v : values) bst.insert(v);
        printList("Inorder after inserts: ", bst.inorder());

        // Test case 2: Search existing
        System.out.println("Search 40: " + bst.search(40));

        // Test case 3: Search non-existing
        System.out.println("Search 25: " + bst.search(25));

        // Test case 4: Delete leaf (20)
        System.out.println("Delete 20: " + bst.delete(20));
        printList("Inorder after deleting 20: ", bst.inorder());

        // Test case 5: Delete node with one child (30) and two children (50)
        System.out.println("Delete 30: " + bst.delete(30));
        System.out.println("Delete 50: " + bst.delete(50));
        printList("Inorder after deleting 30 and 50: ", bst.inorder());
    }
}

File: dataset/Vanilla - Copy/OPENAI_gpt-5-2025-08-07/Task151.cpp
#include <iostream>
#include <vector>
#include <algorithm>

class Task151 {
private:
    struct Node {
        int key;
        Node* left;
        Node* right;
        Node(int k) : key(k), left(nullptr), right(nullptr) {}
    };
    Node* root = nullptr;

    Node* insertRec(Node* node, int key, bool& inserted) {
        if (!node) {
            inserted = true;
            return new Node(key);
        }
        if (key < node->key) node->left = insertRec(node->left, key, inserted);
        else if (key > node->key) node->right = insertRec(node->right, key, inserted);
        else inserted = false;
        return node;
    }

    Node* deleteRec(Node* node, int key, bool& deleted) {
        if (!node) return nullptr;
        if (key < node->key) node->left = deleteRec(node->left, key, deleted);
        else if (key > node->key) node->right = deleteRec(node->right, key, deleted);
        else {
            deleted = true;
            if (!node->left) {
                Node* r = node->right;
                delete node;
                return r;
            }
            if (!node->right) {
                Node* l = node->left;
                delete node;
                return l;
            }
            Node* succ = node->right;
            while (succ->left) succ = succ->left;
            node->key = succ->key;
            node->right = deleteRec(node->right, succ->key, deleted = true);
        }
        return node;
    }

    void inorderRec(Node* node, std::vector<int>& out) const {
        if (!node) return;
        inorderRec(node->left, out);
        out.push_back(node->key);
        inorderRec(node->right, out);
    }

    void freeRec(Node* node) {
        if (!node) return;
        freeRec(node->left);
        freeRec(node->right);
        delete node;
    }

public:
    ~Task151() { freeRec(root); }

    bool search(int key) const {
        Node* cur = root;
        while (cur) {
            if (key == cur->key) return true;
            if (key < cur->key) cur = cur->left; else cur = cur->right;
        }
        return false;
    }

    bool insert(int key) {
        bool inserted = false;
        root = insertRec(root, key, inserted);
        return inserted;
    }

    bool deleteKey(int key) {
        bool deleted = false;
        root = deleteRec(root, key, deleted);
        return deleted;
    }

    std::vector<int> inorder() const {
        std::vector<int> out;
        inorderRec(root, out);
        return out;
    }
};

static void printVec(const std::vector<int>& v) {
    std::cout << "[";
    for (size_t i = 0; i < v.size(); ++i) {
        if (i) std::cout << ", ";
        std::cout << v[i];
    }
    std::cout << "]\n";
}

int main() {
    Task151 bst;
    // Test 1: Insert values and print inorder
    int vals[] = {5, 3, 7, 2, 4, 6, 8};
    for (int v : vals) std::cout << (bst.insert(v) ? "1" : "0") << " ";
    std::cout << "\n";
    printVec(bst.inorder());

    // Test 2: Search existing value
    std::cout << (bst.search(4) ? "1" : "0") << "\n";

    // Test 3: Search non-existing value
    std::cout << (bst.search(10) ? "1" : "0") << "\n";

    // Test 4: Delete leaf (2) and node with one child (3), then print inorder
    std::cout << (bst.deleteKey(2) ? "1" : "0") << "\n";
    std::cout << (bst.deleteKey(3) ? "1" : "0") << "\n";
    printVec(bst.inorder());

    // Test 5: Delete node with two children (7), try duplicate insert (5), delete non-existent (42)
    std::cout << (bst.deleteKey(7) ? "1" : "0") << "\n";
    std::cout << (bst.insert(5) ? "1" : "0") << "\n";
    std::cout << (bst.deleteKey(42) ? "1" : "0") << "\n";
    printVec(bst.inorder());
    return 0;
}